### PR TITLE
feat(NODE-4684)!: remove collection insert, update, remove methods

### PR DESCRIPTION
**Pull request by nbbeeken**
#### What is changing?

Removes the legacy crud methods.

#### What is the motivation for this change?

Modernize code base, remove duplicate code.

### Double check the following

TODO:
- [x] Migration guide
- [x] Feedback on shimming old methods
---
- [ ] Ran `npm run check:lint` script
- [ ] Self-review completed using the [steps outlined here](https://github.com/mongodb/node-mongodb-native/blob/HEAD/CONTRIBUTING.md#reviewer-guidelines)
- [ ] PR title follows the [correct format](https://www.conventionalcommits.org/en/v1.0.0/): `type(NODE-xxxx)[!]: description`
  - Example: `feat(NODE-1234)!: rewriting everything in coffeescript`
- [ ] Changes are covered by tests
- [ ] New TODOs have a related JIRA ticket

### CODE REVIEW

**Comment by nbbeeken** [comment]
I collected the [usage counts of the legacy methods in this patch](https://spruce.mongodb.com/version/63b48ec0a4cf471037033870/tasks?sorts=STATUS%3AASC%3BBASE_STATUS%3ADESC). 

4.0-replica_set reports: `{ insert: 326, update: 19, remove: 9 }` so I can probably look for the removes and maybe the updates, but the inserts are many 😅

**Comment by nbbeeken** [comment]
From slack: Looking into search and replace solutions, there wasn't an approach that would make a significant enough impact in the number of insert usages. We'll defer the work to clean this up to a later time

File: test/integration/crud/crud_api.test.ts [hunk 1]
```diff
@@ -655,71 +655,6 @@ describe('CRUD API', function () {
     }
   });
 
-  it('should correctly execute remove methods using crud api', {
```
Comment (nbbeeken) [inline test/integration/crud/crud_api.test.ts:658]: Wasn't using .remove even and is covered by CRUD spec tests

File: test/integration/crud/find.test.js [hunk 2]
```diff
@@ -1,13 +1,18 @@
...
+const { assert: test } = require('../shared');
 const { expect } = require('chai');
 const sinon = require('sinon');
 const { setTimeout } = require('timers');
 const { Code, ObjectId, Long, Binary, ReturnDocument } = require('../../../src');
 
 describe('Find', function () {
-  before(function () {
-    return setupDatabase(this.configuration);
+  let client;
```
Comment (nbbeeken) [inline test/integration/crud/find.test.js:9]: introducing mocha hooks cus there was a close leak, classic callback didn't reach the end stuff

File: test/integration/crud/find.test.js [hunk 3]
```diff
@@ -926,44 +943,41 @@ describe('Find', function () {
...
-                title: 'Tobi',
-                author: 'Brian',
-                newTitle: 'Woot',
-                meta: { visitors: 0 }
-              },
-              configuration.writeConcernMax(),
-              function (err, r) {
-                // Fetch the id
-                var id = r.insertedIds[0];
+      var db = client.db(configuration.db);
```
Comment (nbbeeken) [inline test/integration/crud/find.test.js:946]: leak fix

File: test/integration/crud/find.test.js [hunk 4]
```diff
@@ -1458,58 +1472,55 @@ describe('Find', function () {
...
-        transactions.push(transaction);
-        // Wrapping object
-        var wrapingObject = {
-          funds: {
-            remaining: 100.5
-          },
-
-          transactions: transactions
-        };
+      var db = client.db(configuration.db);
```
Comment (nbbeeken) [inline test/integration/crud/find.test.js:1475]: leak, etc. I'll stop commenting each one

File: test/integration/node-specific/auto_connect.test.ts [hunk 5]
```diff
@@ -483,15 +483,6 @@ describe('When executing an operation for the first time', () => {
       });
     });
 
-    describe(`#insert()`, () => {
```
Comment (nbbeeken) [inline test/integration/node-specific/auto_connect.test.ts:486]: Don't need to check auto connect on APIs that don't exist

File: test/integration/node-specific/operation_examples.test.ts [hunk 6]
```diff
@@ -1802,77 +1802,6 @@ describe('Operations', function () {
     }
   });
 
-  /**
-   * Example of using keepGoing to allow batch insert using a Promise to complete even when there are illegal documents in the batch
-   *
-   * example-class Collection
-   * example-method insert
-   */
-  it('Should correctly execute insert with keepGoing option on mongod >= 1.9.1 With Promises', {
```
Comment (nbbeeken) [inline test/integration/node-specific/operation_examples.test.ts:1811]: keepGoing removed

File: test/integration/objectid.test.js [hunk 7]
```diff
@@ -1,92 +1,43 @@
...
-        expect(id.toHexString().length).to.equal(24);
-
-        // Locate the first document inserted
-        collection.findOne(id, function (err, document) {
-          expect(err).to.not.exist;
-          expect(id.toHexString()).to.equal(document._id.toHexString());
-          number_of_tests_done++;
-        });
-      });
+  it('generates new ObjectId for documents without _id property', async function () {
```
Comment (nbbeeken) [inline test/integration/objectid.test.js:17]: It failed at first b/c of the insert usage, but then I also cleaned it up cus the setInterval logic was too interesting

File: test/integration/crud/remove.test.js [hunk 8]
```diff
@@ -107,6 +103,21 @@ describe('Remove', function () {
     }
   });
 
+  it('deleteOne', async () => {
```
Comment (baileympearson) [inline test/integration/crud/remove.test.js:106]: it's not clear from the description or the assertions what is under test here.  Also, why did we decide to add this test?  We have unified tests that cover all the assertions here (although implicitly)

File: test/integration/crud/find.test.js (hunk 2 above)
Comment (baileympearson) [reply test/integration/crud/find.test.js:9]: Why did these leaks only show up now?  We didn't change any codepaths so I would expect no change in test behavior.

File: etc/notes/CHANGES_5.0.0.md [hunk 9]
```diff
@@ -124,3 +124,24 @@ cursor.closed // true
...
+```ts
+// Single document insert:
+await collection.insert({ name: 'spot' });
+// Migration:
+await collection.insertMany([{ name: 'spot' }]);
+
+// Multi-document insert:
+await collection.insert([{ name: 'fido' }, { name: 'luna' }])
+// Migration:
+await collection.insertMany([{ name: 'fido' }, { name: 'luna' }])
```
Comment (baileympearson) [inline etc/notes/CHANGES_5.0.0.md:141]: just a thought - a table that has columns for legacy method / existing crud replacement might be an easier format for users to understand.

**Review by baileympearson — CHANGES_REQUESTED** [review]

File: test/integration/collection-management/collection.test.ts [hunk 10]
```diff
@@ -235,7 +235,7 @@ describe('Collection', function () {
           expect(err).to.not.exist;
           collection.insertOne({ c: 1 }, configuration.writeConcernMax(), err => {
             expect(err).to.not.exist;
-            collection.remove({ a: 1 }, configuration.writeConcernMax(), err => {
+            collection.deleteMany({ a: 1 }, configuration.writeConcernMax(), err => {
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply test/integration/collection-management/collection.test.ts]: No I don't think so, CRUD spec and generally we've got [coverage here](https://mciuploads.s3.amazonaws.com/mongo-node-driver/83bce00e28134b62f53b96a96a3d2fbaa2832a92/63b8715030661560b4b2996b/coverage/src/collection.ts.html). Removing

File: test/integration/crud/remove.test.js (hunk 8 above)
Comment (nbbeeken) [reply test/integration/crud/remove.test.js]: Agreed, [spec tests](https://github.com/mongodb/node-mongodb-native/blob/NODE-4684-rm-coll-insert-update-remove/test/spec/crud/unified/deleteMany-comment.json#L68) do check the "deletedCount" so no need for a dupe here.

File: test/tools/runner/hooks/legacy.ts [hunk 11]
```diff
@@ -0,0 +1,34 @@
+import { Collection } from '../../../../src';
```
[the remark this answers is not in the document]
Comment (nbbeeken) [reply test/tools/runner/hooks/legacy.ts]: done

File: etc/notes/CHANGES_5.0.0.md (hunk 9 above)
Comment (nbbeeken) [reply etc/notes/CHANGES_5.0.0.md:141]: done!

Comment (baileympearson) [reply etc/notes/CHANGES_5.0.0.md:141]: Ah, I was thinking something along the lines of 

- replacing 129-130 with something like "Three legacy crud helpers on the collection class have been removed:"
- removing the typescript code example
- putting code examples of before / after in the table, so users clearly see a mapping between the "old code" and "new code"

File: test/integration/crud/find.test.js [hunk 12]
```diff
@@ -786,83 +791,96 @@ describe('Find', function () {
...
-            // Let's modify the document in place
-            collection.findOneAndUpdate(
-              { a: 1 },
-              { $set: { b: 3 } },
-              { returnDocument: ReturnDocument.AFTER },
-              function (err, updated_doc) {
-                test.equal(1, updated_doc.value.a);
-                test.equal(3, updated_doc.value.b);
+            // Test return new document on change
+            collection.insert({ a: 1, b: 2 }, configuration.writeConcernMax(), function (err) {
```
Comment (W-A-James) [inline test/integration/crud/find.test.js:802]: Since we're trying to get rid of instances of `collection.insert`, why don't we get rid of these instances when updating old tests as part of our drive-by cleanup?

**Review by W-A-James — COMMENTED** [review]
Apart from the question I had about replacing instances of `colllection.insert` in tests that are already being modified, lgtm.

File: test/integration/crud/find.test.js (hunk 12 above)
Comment (nbbeeken) [reply test/integration/crud/find.test.js:802]: The change to this test was caused by unindenting the connect callback to fix a leak that showed up when it was erroring. There's a lot more cleanup that could be done here, remove callbacks, remove the old topology filtering, remove the combination of operations (dropColl, createColl, insert should all not be included in a test focused on findOneAndUpdate) and even fix the test name from being all camelcase. But there's only so much drive-by you can do before you're overloading the original task. 

Another reason I leave all of above mentioned legacy practices behind is that it earmarks the test for those of us who've been working on the code for awhile. When this test causes the next engineer some trouble, they'll know it isn't a recent addition by all these other outstanding issues with it. My drive-by goal was to address the leak issue, the next one may be to address any of the above issues when it causes trouble again.

File: etc/notes/CHANGES_5.0.0.md (hunk 9 above)
Comment (nbbeeken) [reply etc/notes/CHANGES_5.0.0.md:141]: I had code samples when I first wrote the table but it isn't easy to read b/c of how wide it gets, it doesn't get the same scroll effect normal code blocks do. Let me fix up the text part and maybe move the code sample below the table

Comment (nbbeeken) [reply etc/notes/CHANGES_5.0.0.md]: Pushed the change, also the bracket are bit hard to notice, "arrayOfDocuments" is obviously just a variable name that implies something. Maybe I should use type annotations instead?

File: etc/notes/CHANGES_5.0.0.md [hunk 13]
```diff
@@ -124,3 +124,33 @@ cursor.closed // true
...
+Three legacy operation helpers on the collection class have been removed:
+
+| Removed API                                    | API to migrate to                                  |
+|------------------------------------------------|----------------------------------------------------|
+| `insert(document)`                             | `insertOne(document)`                              |
+| `insert(arrayOfDocuments)`                     | `insertMany(arrayOfDocuments)`                     |
+| `update(filter)`                               | `updateMany(filter)`                               |
+| `remove(filter)`                               | `deleteMany(filter)`                               |
+
+The `insert` method could accept arrays or single documents, instead for single document inserts `insertOne` or an `insertMany` call with an array argument of one should be used.
```
Comment (baileympearson) [inline etc/notes/CHANGES_5.0.0.md:139]:
```suggestion
The `insert` method accepted an array of documents for multi-document inserts and a single document for single document inserts.  `insertOne` should now be used for single-document inserts and `insertMany` should be used for multi-document inserts.
```

**Review by baileympearson — APPROVED** [review]
